Enhanced the 'Dense' layer to support batch matrix multiplication.

eaplatanios opened the pull request:
This enhances the `Dense` layer to be more aligned with the `matmul` operation. More specifically, it allows it to use the batch variant of `matmul` if the rank of `weight` is equal to `3`. Not sure if we want to merge this here but I use it often in my work and it also aligns `Dense` with the functionality of `matmul`.

rxwei commented on Sources/TensorFlow/Layers/Core.swift:
```diff
@@ -166,6 +171,10 @@ public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
     /// - Returns: The output.
     @differentiable
     public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        if weight.rank == 3 {
```
@bgogul Is there a lazy tensor concern here?

eaplatanios replied:
If believe there is, but what if I write this as: `weight.rankTensor == Tensor<Int>(3)`?

rxwei replied:
Yeah, that might work a lot better, but I'll wait for @bgogul to confirm.

eaplatanios replied:
Actually, given that `if` statements are not currently being lowered to TF ops, this should definitely affect lazy tensor because the condition depends on `weight`. I can move that condition to the initializer and add a boolean stored property named `batched`. Using that in the `if` statement will be safe with lazy tensor. @rxwei what do you think?

rxwei replied:
Hmm, I'd very much prefer keeping the existing semantics if possible.

eaplatanios replied:
This will retain the same semantics. It just moves the rank check to the initializer.

bgogul replied:
Yes, this would cause materialization in lazy tensor, and thereby, break the trace. Even changing the condition to `weight.rankTensor == Tensor<Int>(3)` won't fix it. Adding the `batched` property would fix it though. 

That said, I don't think it is sustainable to have to rewrite code to help lazy tensor in the long run. If the code becomes too awkward and ugly, just do what makes the most sense. The right way to fix this would be to track shapes in the swift side.

eaplatanios replied:
I just pushed the change so you can see what I mean.

eaplatanios replied:
@bgogul Thanks for the confirmation! I already switched to the `batched` property. And yes, I agree with what you say. In this case I believe it's ok to modify the code given how prevalent dense layers are and also given that lowering if-statements to TF ops may take a while.

rxwei commented on Sources/TensorFlow/Layers/Core.swift:
```diff
@@ -138,6 +138,11 @@ public struct Reshape<Scalar: TensorFlowFloatingPoint>: ParameterlessLayer {
 /// a weight matrix, `bias` is a bias vector, and `activation` is an element-wise activation
 /// function.
+///
+/// This layer also supports 3-dimensional weight tensors with 2-dimensional bias matrices. In this 
```
I think we've been using "n-D" in most API documentation. If so, I suggest changing this. Otherwise never mind.
```suggestion
/// This layer also supports 3-D weight tensors with 2-D bias matrices. In this 
```

rxwei commented on Sources/TensorFlow/Layers/Core.swift:
```diff
@@ -138,6 +138,11 @@ public struct Reshape<Scalar: TensorFlowFloatingPoint>: ParameterlessLayer {
+///
+/// This layer also supports 3-dimensional weight tensors with 2-dimensional bias matrices. In this 
+/// case the first dimension of both is treated as the batch size that is aligned with the first 
+/// dimension of `input` and the batch variant of the `matmul` operation is used, thus using a 
```
```suggestion
/// dimension of `input` and the batch variant of the `matmul(_:_:)` operation is used, thus using a 
```

eaplatanios commented:
Thanks for the feedback Richard! I addressed all comments and now waiting for confirmation from @bgogul that the current approach is "`LazyTensor`-friendly".

eaplatanios commented:
This should be ready for review now. I switched to using the `batched` property.

rxwei commented on Sources/TensorFlow/Layers/Core.swift:
```diff
@@ -158,6 +165,7 @@ public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
         self.weight = weight
```
Should we add preconditions here to assert something like `weight.rank <= 3` and `bias.rank <= 2`?

eaplatanios replied:
Done.
